src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -139,25 +77,72 @@ void mf::XWaylandServer::execl_xwayland(XWaylandSpawner const& spawner, int wl_c
             xwayland_path.c_str(),
             dsp_str.c_str(),
             "-rootless",
-            "-wm", wm_fd_str.c_str(),
+            "-wm", x11_fd_str.c_str(),
             "-terminate",
         };
 
     for (auto const& fd : spawner.socket_fds())
     {
-        XWaylandSpawner::set_cloexec(fd, false);
+        mf::XWaylandSpawner::set_cloexec(fd, false);
         args.push_back("-listen");
         // strdup() may look like a leak, but execvp() will trash all resource management
         args.push_back(strdup(std::to_string(fd).c_str()));
     }
 
-    args.push_back(NULL);
+    args.push_back(nullptr);
 
     execvp(xwayland_path.c_str(), const_cast<char* const*>(args.data()));
 }
 
-namespace
+auto fork_xwayland_process(
+    mf::XWaylandSpawner const& spawner,
+    std::string const& xwayland_path) -> mf::XWaylandServer::XWaylandProcess
 {
+    enum { mir_side, xwayland_side, pipe_size };
+    int wayland_pipe[pipe_size], x11_pipe[pipe_size];
+
+    if (socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, wayland_pipe) < 0)
+    {
+        // "Shouldn't happen" but continuing is weird.
+        BOOST_THROW_EXCEPTION(std::runtime_error("Wayland socketpair failed"));
+    }
+
+    if (socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, x11_pipe) < 0)
+    {
+        // "Shouldn't happen" but continuing is weird.
+        close(wayland_pipe[mir_side]);
+        close(wayland_pipe[xwayland_side]);
+        BOOST_THROW_EXCEPTION(std::runtime_error("X11 socketpair failed"));
+    }
+
+    mir::log_info("Starting XWayland");
+    pid_t const xwayland_pid = fork();
+
+    switch (xwayland_pid)
+    {
+    case -1:
+        close(wayland_pipe[mir_side]);
+        close(wayland_pipe[xwayland_side]);
+        close(x11_pipe[mir_side]);
+        close(x11_pipe[xwayland_side]);
+        BOOST_THROW_EXCEPTION(std::runtime_error("Failed to fork XWayland process"));
+
+    case 0:
+        close(wayland_pipe[mir_side]);
+        close(x11_pipe[mir_side]);
+        exec_xwayland(spawner, xwayland_path, wayland_pipe[xwayland_side], x11_pipe[xwayland_side]);
+        BOOST_THROW_EXCEPTION(std::logic_error("Should be unreachable"));
```

Comment:
This is not actually going to be helpful, is it? This should probably be an `abort()` to tear everything down faster, because we're in a forked process?

Does the main process have a way of noticing this?

Reply:
Done

It seems to just lock up for a while as it waits for xwayland to start. Might look at improving the error path.